docs/custom-component-one-off.qmd

```diff
@@ -0,0 +1,325 @@
+    script={"src": "tableComponent.js", "type": "module"},
+    stylesheet={"href": "tabulator.min.css"},
+    all_files=True,
+)
+
+def output_tabulator(id, height="200px"):
+    return ui.div(
+        tabulator_dep,
+        # Use resolve_id so that our component will work in a module
+        id=resolve_id(id),
+        class_="shiny-tabulator-output",
+        style=f"height: {height}",
+    )
+```
+
+We use the `HTMLDependency` function to bind up the assets needed for tabulator that we made in the previous step and making sure that they're included in our app anytime the `output_tabulator()` function is called (but not more than once.)
+
+:::{.callout-note}
+Note the use of `all_files=True` here. This makes it so we can do the esm import of the tabulator library. Otherwise `tabulator_esm.min.js` would not be hosted and the js library couldn't find it.
+:::
+
+Now, the `output_tabulator()` function can be called anywhere we want to render a table in our app.
+
+## The `render_tabulator()` decorator
+
+Now we've got the client-side logic finished, we need to write a custom render decorator that sends our data into the component.
+
+To do this we can leverage some tools provided by Shiny in the `shiny.render.transformer` subpackage.
+
+```{.python filename="app.py"}
+from shiny.render.transformer import (
+    output_transformer,
+    resolve_value_fn,
+    TransformerMetadata,
+    ValueFn,
+)
+
+
+@output_transformer
+async def render_tabulator(
+    _meta: TransformerMetadata,
+    _fn: ValueFn[pd.DataFrame | None],
+):
+    res = await resolve_value_fn(_fn)
+    if res is None:
+        return None
+
+    if not isinstance(res, pd.DataFrame):
+        # Throw an error if the value is not a dataframe
+        raise TypeError(f"Expected a pandas.DataFrame, got {type(res)}. ")
+
+    # Get data from dataframe as a list of lists where each inner list is a
+    # row, column names as array of strings and types of each column as an
+    # array of strings
+    return {
+        "data": res.values.tolist(),
+        "columns": res.columns.tolist(),
+        "type_hints": res.dtypes.astype(str).tolist(),
+    }
+```
+
+The `output_transformer` decorator is a decorator factory that takes a function that returns a dictionary of data to be passed to the client side. The function that it decorates is passed two arguments: `_meta` and `_fn`.
+
+`_meta` is a dictionary of metadata about the function that is being decorated. We don't use it in our example.
+
+`_fn` is the function that is being decorated. Aka the function that goes below the `@render_tabulator()` in your app's server code. In this case we are expecting that that function returns either a pandas dataframe or `None`.
+
+In the code above we use types so that we can get some type checking in our IDE, but these are not required. Also note that the decorated function is an async function, so we need to use the `await` keyword when we call it for `resolve_value_fn()`.
```

Comment:
I moved some stuff around and added a callout about the async-ness. It's a bit awkward because it's required to allow the user to be uninhibited with their function return type, but in this article, we are the user so in theory that flexibility is not needed.